### ISSUE TITLE
GOkb style Master list

### ISSUE THREAD

**Issue by iyawnis** [k-int/KBPlus#993]
"252 = The ‘master’ list is a package which contains TIPPs for all the TIs known to be supplied by a specific Content Provider - that is, it is a list compiled by combining all the other lists we know of for that provider. Since the same TI will appear in multiple packages for the same Provider, these will need de-duplicating. Earliest start date across all TIPPs used as Start Date on Master package TIPP. Latest end date across all TIPPs used as End Date on Master package TIPP. "

Is this something that will be produced on-demand? E.g A screen where the user will input the provider of the master list, and then an export is created? Or is this an automatic process, that will generate one package for every provider in the system, and keep it updated with a scheduled overnight task? 

And to make sure I got the right idea: We create a new 'master' package for each provider. For every TitleInstance associated with an existing TIPP for that provider, we have a new TIPP keeping any associations and properties same, but Package will be the master list. 
As you said, we might have the same TI on several TIPPs. What are the criteria for deciding which TIPP details to keep ('these will need de-duplicating')?

**Comment by ostephens** [comment]
This is something that needs to be created and updated with a scheduled task.

I think that rather than keeping the details from one of the TIPPs it would be preferable to:
- Use the coverage start date/vol/issue from the TIPP with the earliest start date
- Use the coverage end date/vol/issue from the TIPP with the latest end date

However, it is not clear to me what should happen with other properties - e.g. coverage notes. We may want to look at how GOKb handles this as the brief was to do this in a similar way (but there was some question of whether GOKb could end up with multiple TIPPs for the same TI, and was made clear that this wasn't something that KB+ wanted).

The GOKb code can be tracked back from https://github.com/k-int/gokb-phase1/blob/dev/server/gokb/grails-app/jobs/org/gokb/MasterListUpdateJob.groovy. @sosguthorpe wrote this I think so may be able to advise on any logic used in terms of selecting TIPP properties for the Master list

**Comment by sosguthorpe** [comment]
From memory, I think we go through each package associated with a 
particular Org and copy each TIPP but change the package to the "Master 
package" for the current Org. So all TIPP properties would be duplicated 
except the package.

**Comment by ostephens** [comment]
Thanks @sosguthorpe. Does that mean in GOKb the Master package has multiple TIPPs for the same TI? That is the thing that the KB+ data managers have said they _don't_ want

**Comment by sosguthorpe** [comment]
Had to consult the code because I couldn't remember :)

When a master TIPP is created it is linked to the TIPP on which it is 
based. However, if the current TIPP has no master associated with it we 
check for an existing one, in the correct master package, that has the 
same TI and Platform. If we find one then we update that one and link to 
this TIPP too. That means that the last none master TIPP that has the 
same TI and Platform will be the one used to set the properties on the 
master. Typing this has made me realise that this is actually probably 
not what we need now since the extension of the TIPP properties. We 
should probably be using earliest start date and latest end dates etc.

In short there is one TIPP per TI on the master and that Master TIPP 
could be linked to multiple regular TIPPs. The property settings 
probably need to be revisited now that we have extended what we know 
about the TIPPs.

**Comment by ostephens** [comment]
Thanks @sosguthorpe

**Comment by iyawnis** [comment]
Looking at the code, I think there are no duplicated tipps, line 257 in
PackageService. It looks to me that all of the logic can be transferred
over.

**Comment by ostephens** [comment]
@latusaki based on @sosguthorpe's last comment may need to enhance the logic slightly?

**Comment by iyawnis** [comment]
Yeah, change the behaviour on what gets copied over to the master tipp.

**Comment by ostephens** [comment]
In terms of other fields:

tipp_access_start
tipp_access_end

should follow same logic as tipp_start_date/tipp_end_date (i.e. take the earliest/latest - and null on the tipp_access_end/tipp_end_date always taken as the latest date)

coverage_depth should always be fulltext.

The fields I'm not sure what the rules should be are:

coverage_note
tipp_embargo
tipp_delayedoa
tipp_hybridoa
tipp_payment

I think it was this issue that led us to have links between the master tipp and other tipps in GOKb - so we didn't need to merge this stuff together

**Comment by sosguthorpe** [comment]
My personal opinion would be to favour the less restrictive values (if 
that way of thinking can be applied to the property). So hybridoa a 
value of "Yes" would carry the most weight.

**Comment by iyawnis** [comment]
Does this mean for a master package, all tipps will have the same properties? Same start/end dates access start /end etc? Are there any properties that we need to keep from the original TIPP when creating the master TIPP other than the title instance?

**Comment by ianibbo** [comment]
Ideally, I think we merge dates on the master list to be the earliest
possible start and latest possible end.. Is that right @ostephens?

Ian Ibbotson
Director
Knowledge Integration Ltd
35 Paradise Street, Sheffield. S3 8PZ
T: 0114 273 8271
M: 07968 794 630
W: http://www.k-int.com

**Comment by ostephens** [comment]
Yes @ianibbo  - that's correct - the TIPPs would have the properties based on the TIPPs that were used as the source of the TIPP.

e.g. 

TIPP 1 for TI A in Package B from Provider X
- Start date = 2001-01-01
- End date = blank

TIPP 2  for TI A  in Package C from Provider X
- Start date = 2010-01-01
- End date = 2013-12-31

Master TIPP for TI A for Provider X
- Start date = 2001-01-01
- End date = blank

Same logic for Access Start/Access End dates

Based on the comment from @sosguthorpe we need a logic telling you which values of:

tipp_embargo
tipp_delayedoa
tipp_hybridoa
tipp_payment

are seen as the 'overriding' values (e..g if tipp_payment = free on one TIPP and 'paid' on another, should the master TIPP have tipp_payment = free, or tip_payment = paid...). Some of these I think might work (e.g. some value > blank/unspecified) but I'm not sure we can make an automated decision in all scenarios. We could potentially add a 'mixed' (or similar) value here to indicate that the contributing TIPPs disagree...

**Comment by ostephens** [comment]
Picking up a Q from @latusaki via chat and posting here for the record:

**Q: What Package start / end dates should be used for the auto-generated master?**

A: Master packages that have been manually created on the system by the data managers are created with 'end dates' set in the far future so they are essentially always current.

We should take the same approach with the auto-generated master packages. The 'start date' should be in the far past (e.g. 0001-01-01) and the end date in the far future (e.g. 9999-12-31)

**Comment by iyawnis** [comment]
I have the following method inside the master list service, we just need to add rules for everything else..

``` groovy
  /**
  * Move all attributes that we need to keep from tipp_b to tipp_a 
  **/
  private void processDuplicatedTipps(tipp_a, tipp_b){
    //In here we compare the properties of the two tipps, and we transfer the ones we want to keep to tipp_a
    if(tipp_a.startDate < tipp_b.startDate) tipp_a.startDate = tipp_b.startDate;
    if(tipp_a.endDate > tipp_b.endDate) tipp_a.endDate = tipp_b.endDate;
    if(tipp_a.startVolume < tipp_b.startVolume) tipp_a.startVolume = tipp_b.startVolume;
    if(tipp_a.endVolume < tipp_b.endVolume) tipp_a.endVolume = tipp_b.endVolume;
    if(tipp_a.endIssue < tipp_b.endIssue) tipp_a.endIssue = tipp_b.endIssue;
    if(tipp_a.startIssue < tipp_b.startIssue) tipp_a.startIssue = tipp_b.startIssue;

    if(!tipp_a.hostPlatformURL) tipp_a.hostPlatformURL = tipp_b.hostPlatformURL;


    //mark this tipp title null so we don't try process it again
    tipp_b.title = null

  }
```

**Comment by iyawnis** [comment]
```
2015-08-12 15:04:04,183 [http-bio-8080-exec-3] DEBUG kbplus.PackageService  - Finished updating master package 1593
2015-08-12 15:04:04,232 [http-bio-8080-exec-2] ERROR util.JDBCExceptionReporter  - Lock wait timeout exceeded; try restarting transaction
2015-08-12 15:04:04,666 [http-bio-8080-exec-2] ERROR errors.GrailsExceptionResolver  - SQLException occurred when processing request: [GET] /demo/admin/pkgService
Lock wait timeout exceeded; try restarting transaction. Stacktrace follows:
java.sql.SQLException: Lock wait timeout exceeded; try restarting transaction
    at com.mysql.jdbc.SQLError.createSQLException(SQLError.java:1084)
    at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:4232)
    at com.mysql.jdbc.MysqlIO.checkErrorPacket(MysqlIO.java:4164)
    at com.mysql.jdbc.MysqlIO.sendCommand(MysqlIO.java:2615)
    at com.mysql.jdbc.MysqlIO.sqlQueryDirect(MysqlIO.java:2776)
    at com.mysql.jdbc.ConnectionImpl.execSQL(ConnectionImpl.java:2838)
    at com.mysql.jdbc.PreparedStatement.executeInternal(PreparedStatement.java:2082)
    at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:2334)
    at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:2262)
    at com.mysql.jdbc.PreparedStatement.executeUpdate(PreparedStatement.java:2246)
    at net.bull.javamelody.JdbcWrapper.doExecute(JdbcWrapper.java:403)
    at net.bull.javamelody.JdbcWrapper$StatementInvocationHandler.invoke(JdbcWrapper.java:128)
    at net.bull.javamelody.JdbcWrapper$DelegatingInvocationHandler.invoke(JdbcWrapper.java:285)
    at com.k_int.kbplus.PackageService.updateMasterFor(PackageService.groovy:111)
    at GrailsMelodyGrailsPlugin$_closure4_closure16_closure17.doCall(GrailsMelodyGrailsPlugin.groovy:184)
    at com.k_int.kbplus.PackageService$_updateAllMasters_closure1.doCall(PackageService.groovy:32)
    at com.k_int.kbplus.PackageService.updateAllMasters(PackageService.groovy:31)
    at GrailsMelodyGrailsPlugin$_closure4_closure16_closure17.doCall(GrailsMelodyGrailsPlugin.groovy:184)
    at com.k_int.kbplus.AdminController.pkgService(AdminController.groovy:34)
    at grails.plugin.cache.web.filter.PageFragmentCachingFilter.doFilter(PageFragmentCachingFilter.java:198)
    at grails.plugin.cache.web.filter.AbstractFilter.doFilter(AbstractFilter.java:63)
    at net.bull.javamelody.JspWrapper.invoke(JspWrapper.java:150)
    at net.bull.javamelody.JdbcWrapper$DelegatingInvocationHandler.invoke(JdbcWrapper.java:285)
    at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:198)
    at net.bull.javamelody.MonitoringFilter.doFilter(MonitoringFilter.java:176)
    at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145)
    at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615)
    at java.lang.Thread.run(Thread.java:745)
```

**Comment by iyawnis** [comment]
```
2015-08-12 15:12:31,959 [http-bio-8080-exec-8] DEBUG kbplus.PackageService  - Finished updating master package 2228
2015-08-12 15:12:32,021 [http-bio-8080-exec-9] ERROR events.PatchedDefaultFlushEventListener  - Could not synchronize database state with session
org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [com.k_int.kbplus.TitleInstancePackagePlatform#188932]
    at com.k_int.kbplus.PackageService.setOrUpdateMasterTippFor(PackageService.groovy:275)

```

**Comment by iyawnis** [comment]
@ostephens  I believe there is no reason to associate a master tipp with a TIPP now? Since we only process one tipp when we have duplication, the master would only be associated with one of them

**Comment by ostephens** [comment]
Yes - the Master TIPP is just another TIPP which is associated with the master package - no need for it to link to any other TIPPs

**Comment by iyawnis** [comment]
2015-08-12 17:00:44,754 [http-bio-8080-exec-4] ERROR errors.GrailsExceptionResolver  - StaleObjectStateException occurred when processing request: [GET] /demo/admin/pkgService
Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [com.k_int.kbplus.Package#1590]. Stacktrace follows:
org.hibernate.StaleObjectStateException: Row was updated or deleted by another transaction (or unsaved-value mapping was incorrect): [com.k_int.kbplus.Package#1590]
    at com.k_int.kbplus.PackageService$$EPLAB4vo.updateMasterFor(PackageService.groovy:111)

**Comment by iyawnis** [comment]
Ok. I am still getting errors sometimes, trying to figure out the cause..

**Comment by iyawnis** [comment]
@ostephens  need a list of rules for the duplicated data, so that we can push this to test. Here are the current rules:

``` groovy
  /**
  * Move all attributes that we need to keep from tipp_b to tipp_a 
  **/
  private void processDuplicatedTipps(tipp_a, tipp_b){
    //In here we compare the properties of the two tipps, and we transfer the ones we want to keep to tipp_a
    if(tipp_a.startDate < tipp_b.startDate) tipp_a.startDate = tipp_b.startDate;
    if(tipp_a.endDate > tipp_b.endDate) tipp_a.endDate = tipp_b.endDate;
    if(tipp_a.startVolume < tipp_b.startVolume) tipp_a.startVolume = tipp_b.startVolume;
    if(tipp_a.endVolume < tipp_b.endVolume) tipp_a.endVolume = tipp_b.endVolume;
    if(tipp_a.endIssue < tipp_b.endIssue) tipp_a.endIssue = tipp_b.endIssue;
    if(tipp_a.startIssue < tipp_b.startIssue) tipp_a.startIssue = tipp_b.startIssue;

    if(!tipp_a.hostPlatformURL) tipp_a.hostPlatformURL = tipp_b.hostPlatformURL;


    //mark this tipp title null so we don't try process it again
    tipp_b.title = null

  }
```
